lightning-persister/src/util.rs

```diff
@@ -40,21 +33,23 @@ fn path_to_windows_str<T: AsRef<OsStr>>(path: T) -> Vec<winapi::shared::ntdef::W
 }
 
 #[allow(bare_trait_objects)]
-pub(crate) fn write_to_file<D: DiskWriteable>(path: PathBuf, filename: String, data: &D) -> std::io::Result<()> {
-	fs::create_dir_all(path.clone())?;
+pub(crate) fn write_to_file<W: Writeable>(filename_with_path: String, data: &W) -> std::io::Result<()> {
```

Comment:
Not sure I understand why `PathBuf` is avoided so much. Can't we simply change the the type of `filename_with_path` to `PathBuf` and never convert it to `String`? Seems we are allocating by using `format!`, which has the double separator drawback, too.

Reply:
The problem is that the KVStorePersister trait uses a `String` key instead of a `PathBuf` because it's intended to be used for more than filesystem persistence.  KVStorePersister is the caller of `write_to_file` so it only has a `String` key.  

I think the idea was that if we are going to need to convert the PathBuf to a String in order to persist via KVStorePersister then we might as well avoid building the PathBuf in the first place.

As for this method, I also just realized we were recreating the PathBuf and cloning filename_from_path down where we do the rename/copy from tmp to final.  I removed it and just re-used the PathBuf we create at the top.

Reply:
> The problem is that the KVStorePersister trait uses a `String` key instead of a `PathBuf` because it's intended to be used for more than filesystem persistence. KVStorePersister is the caller of `write_to_file` so it only has a `String` key.
> 

Doesn't `PathBuf::push` just store a reference? Its parameter is `P: AsRef<Path>`, where `str` and `String`'s implementations call `Path::new`, which is described as "a cost-free conversion", and where there is a blanket implementation for reference types.

> I think the idea was that if we are going to need to convert the PathBuf to a String in order to persist via KVStorePersister then we might as well avoid building the PathBuf in the first place.
> 

Not sure I follow why `FilesystemPersister::persist`'s implementation of `KVStorePersister` would need to convert to a string. It's given a string and from there it can cheaply wrap it in a `PathBuf` and never need to convert back to a string if `write_to_file` is changed to take a `PathBuf`.

Reply:
Yeah it's not the implementation of KVStorePersister that has the conversion issue (going from String to PathBuf is free I think) it's the usage of `persist`.  On `main` it uses PathBuf's for building up the paths for manager,graph,monitors but `persist` needs the paths as &str so originally this branch was taking the PathBuf's and converting to str

in the next iteration I removed the usage of PathBuf for building up the paths in the first place.

Reply:
> Yeah it's not the implementation of KVStorePersister that has the conversion issue (going from String to PathBuf is free I think) it's the usage of `persist`. On `main` it uses PathBuf's for building up the paths for manager,graph,monitors but `persist` needs the paths as &str so originally this branch was taking the PathBuf's and converting to str
> 

Do you mean for `FilesystemPersister::path_to_channel_data`? Isn't that a one-time occurrence? `persist` takes the key, yes, but it's the `FilesystemPersister` implementation that tacks on `path_to_channel_data` as the path prefix.

Reply:
Hm I was just trying to explain how we got here, since yeah the original provided implementation was only part of FilesystemPersister and it used PathBuf for everything (manager, graph, monitor).  

At this point this PR is such a mess it's getting hard to follow for me.  I was mostly following matts comments that led us here.  Can we maybe just sync on this on slack real quick to make a final call so we can just get this over the finish line?